### PR TITLE
Emit non BMP chars as XObjects in PDF

### LINKED ISSUE #20616: Type 42 chars beyond BMP not displayed in PDF
### Bug report

**Bug summary**

Characters from beyond the Basic Multilingual Plane (BMP) in Unicode are not displayed in PDFs with Type 42 fonts. Characters beyond the BMP have a code point greater than 65535 and cannot be encoded in a fixed-size 2-byte encoding, such as the obsolete UCS-2. My understanding is that the CID maps still use such an encoding and cannot handle code points beyond that.
 
In the case of the `m` in STIX Sans, it's implemented via a virtual font embedded in the base font with chars shifted to higher code points, here code point 120366.

**Code for reproduction**
```python
import matplotlib.pyplot as plt
from matplotlib import rcParams

rcParams["pdf.fonttype"] = 42
rcParams["mathtext.fontset"] = "stixsans"

plt.text(0.5, 0.5, "Mass $m$ \U00010308")
plt.savefig("beyond_bmp.pdf")
```

**Actual outcome**

beyond_bmp_pdf
[beyond_bmp.pdf](https://github.com/matplotlib/matplotlib/files/6792729/beyond_bmp.pdf)

**Expected outcome**
beyond_bmp

**Possible solutions**
We could take a similar approach as we did for Type 3 fonts. There any char>255 is embedded via an `XObject` and not via the font directly. So, here the solution would be to use `XObjects` if the char>65535 in text and math mode.

I have a local, modified version of matplotlib based on https://github.com/matplotlib/matplotlib/pull/20615 which implements this approach. It extends the use of `_font_supports_char` and restricts the supported range for Type 42 to <=65535. I can polish this local fix into a PR if people what to go this route.

**Matplotlib version**

  * Operating system: Debian 11
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`):  3.4.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg (but I assume pdf)
  * Python version: 3.9.2
  * Matplotlib installed with pip in a virtual env

### PR DESCRIPTION
**Pull request by sauerburger**
## PR Summary

Closes #20616

This PR address https://github.com/matplotlib/matplotlib/issues/20616 and emits Type 42 chars from beyond the Basic Multilingual Plane as XObjects, similar to the treatment of multi-byte chars in Type 3 fonts.

This PR builds on top of https://github.com/matplotlib/matplotlib/pull/20615.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Review by aitikgupta — APPROVED** [review]
I think this is close to what @anntzer mentioned by splitting characters into 'multi-bytes' for Type 42 as well (briefly during the call and on https://github.com/matplotlib/matplotlib/issues/18191).

Looks good! 🎉

File: lib/matplotlib/backends/backend_pdf.py [hunk 1]
```diff
@@ -321,6 +321,21 @@ def pdfRepr(obj):
                         .format(type(obj)))
 
 
+def _font_supports_char(fonttype, char):
+    """
+    Returns True if the font is able to provided the char in a PDF
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:326]:
```suggestion
    Returns True if the font is able to provide *char* in a PDF.
```

File: lib/matplotlib/backends/backend_pdf.py [hunk 2]
```diff
@@ -1194,6 +1209,8 @@ def embedTTFType42(font, characters, descriptor):
             wObject = self.reserveObject('Type 0 widths')
             toUnicodeMapObject = self.reserveObject('ToUnicode map')
 
+            bbox = [cvt(x, nearest=False) for x in font.bbox]
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py:1212]: Does this need to be calculated so early, and not right before emitting the XObjects?

File: lib/matplotlib/backends/backend_pdf.py (hunk 1 above)
Comment (sauerburger) [reply lib/matplotlib/backends/backend_pdf.py]: Done

File: lib/matplotlib/backends/backend_pdf.py (hunk 2 above)
Comment (sauerburger) [reply lib/matplotlib/backends/backend_pdf.py]: No, this is how it was done for Type 3. I've moved it down to the point where the XObjects are emitted.

**Review by jklymak — APPROVED** [review]
I'm approving in lieu of @aitikgupta, not because I understand the font handling here.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**